In [1]:
import numpy as np
import pandas as pd 
import matplotlib.pyplot as plt
import mlxtend

from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules

#Association rules for a clothing company

In [2]:
dataset=[['Skirt', 'Sneakers', 'Scarf', 'Pants', 'Hat'],
        ['Sunglasses', 'Skirt', 'Sneakers', 'Pants', 'Hat'],
        ['Dress', 'Sandals', 'Scarf', 'Pants', 'Heels'],
        ['Dress', 'Necklace', 'Earrings', 'Scarf', 'Hat', 'Heels', 'Hat'],
        ['Earrings', 'Skirt', 'Skirt', 'Scarf', 'Shirt', 'Pants']]

In [3]:
te=TransactionEncoder()
te_ary=te.fit(dataset).transform(dataset)    
data=pd.DataFrame(te_ary, columns=te.columns_)  
data

Dress  Earrings    Hat  Heels  ...  Shirt  Skirt  Sneakers  Sunglasses
0  False     False   True  False  ...  False   True      True       False
1  False     False   True  False  ...  False   True      True        True
2   True     False  False   True  ...  False  False     False       False
3   True      True   True   True  ...  False  False     False       False
4  False      True  False  False  ...   True   True     False       False

[5 rows x 12 columns]

In [4]:
data.columns

Index(['Dress', 'Earrings', 'Hat', 'Heels', 'Necklace', 'Pants', 'Sandals',
       'Scarf', 'Shirt', 'Skirt', 'Sneakers', 'Sunglasses'],
      dtype='object')

In [5]:
frequent_itemsets=apriori(data, min_support=0.6, use_colnames=True) 
frequent_itemsets

support        itemsets
0      0.6           (Hat)
1      0.8         (Pants)
2      0.8         (Scarf)
3      0.6         (Skirt)
4      0.6  (Scarf, Pants)
5      0.6  (Pants, Skirt)

In [6]:
association_rules(frequent_itemsets,metric="lift",min_threshold=0.7)

antecedents consequents  antecedent support  ...    lift  leverage  conviction
0     (Scarf)     (Pants)                 0.8  ...  0.9375     -0.04         0.8
1     (Pants)     (Scarf)                 0.8  ...  0.9375     -0.04         0.8
2     (Pants)     (Skirt)                 0.8  ...  1.2500      0.12         1.6
3     (Skirt)     (Pants)                 0.6  ...  1.2500      0.12         inf

[4 rows x 9 columns]

There is a high probability for customers to buy Skirts and Pants together(lift = 1.25 > 1)so these 2 items should be put together.

#Market Basket Optimisation

In [7]:
df=pd.read_csv('/content/Market_Basket_Optimisation.csv',header=None)
df.head()

0          1           2   ...               17       18         19
0         shrimp    almonds     avocado  ...  frozen smoothie  spinach  olive oil
1        burgers  meatballs        eggs  ...              NaN      NaN        NaN
2        chutney        NaN         NaN  ...              NaN      NaN        NaN
3         turkey    avocado         NaN  ...              NaN      NaN        NaN
4  mineral water       milk  energy bar  ...              NaN      NaN        NaN

[5 rows x 20 columns]

In [8]:
df.shape

(7501, 20)

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7501 entries, 0 to 7500
Data columns (total 20 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   0       7501 non-null   object
 1   1       5747 non-null   object
 2   2       4389 non-null   object
 3   3       3345 non-null   object
 4   4       2529 non-null   object
 5   5       1864 non-null   object
 6   6       1369 non-null   object
 7   7       981 non-null    object
 8   8       654 non-null    object
 9   9       395 non-null    object
 10  10      256 non-null    object
 11  11      154 non-null    object
 12  12      87 non-null     object
 13  13      47 non-null     object
 14  14      25 non-null     object
 15  15      8 non-null      object
 16  16      4 non-null      object
 17  17      4 non-null      object
 18  18      3 non-null      object
 19  19      1 non-null      object
dtypes: object(20)
memory usage: 1.1+ MB


We have to convert the DataFrame into list of lists to do the transaction encoding

In [10]:
lists = []
for i in range (0, 7501):
    lists.append([str(df.values[i,j]) for j in range(0, 20)])

In [11]:
TE = TransactionEncoder()
array = TE.fit(lists).transform(lists)
T_df = pd.DataFrame(array, columns = TE.columns_)
T_df

asparagus  almonds  antioxydant juice  ...   yams  yogurt cake  zucchini
0          False     True               True  ...   True        False     False
1          False    False              False  ...  False        False     False
2          False    False              False  ...  False        False     False
3          False    False              False  ...  False        False     False
4          False    False              False  ...  False        False     False
...          ...      ...                ...  ...    ...          ...       ...
7496       False    False              False  ...  False        False     False
7497       False    False              False  ...  False        False     False
7498       False    False              False  ...  False        False     False
7499       False    False              False  ...  False        False     False
7500       False    False              False  ...  False         True     False

[7501 rows x 121 columns]

We should not forget that the nan value was also encoded so we have to drop the column nan from DataFrame T_df.

In [12]:
T_df.drop(['nan'],axis=1,inplace=True)

In [13]:
a_rules = apriori(T_df, min_support = 0.04, use_colnames = True)
a_rules

support                      itemsets
0   0.087188                     (burgers)
1   0.081056                        (cake)
2   0.046794                   (champagne)
3   0.059992                     (chicken)
4   0.163845                   (chocolate)
5   0.080389                     (cookies)
6   0.051060                 (cooking oil)
7   0.179709                        (eggs)
8   0.079323                    (escalope)
9   0.170911                (french fries)
10  0.043061                 (fresh bread)
11  0.063325             (frozen smoothie)
12  0.095321           (frozen vegetables)
13  0.052393               (grated cheese)
14  0.132116                   (green tea)
15  0.098254                 (ground beef)
16  0.049460               (herb & pepper)
17  0.047460                       (honey)
18  0.076523              (low fat yogurt)
19  0.129583                        (milk)
20  0.238368               (mineral water)
21  0.065858                   (olive oil)
22  0.095054                    (pancakes)
23  0.042528                      (salmon)
24  0.071457                      (shrimp)
25  0.050527                        (soup)
26  0.174110                   (spaghetti)
27  0.068391                    (tomatoes)
28  0.062525                      (turkey)
29  0.058526            (whole wheat rice)
30  0.052660    (mineral water, chocolate)
31  0.050927         (mineral water, eggs)
32  0.040928  (mineral water, ground beef)
33  0.047994         (mineral water, milk)
34  0.059725    (mineral water, spaghetti)

In [14]:
association_rules(a_rules, metric = 'lift', min_threshold = 1)

antecedents      consequents  ...  leverage  conviction
0  (mineral water)      (chocolate)  ...  0.013604    1.073256
1      (chocolate)  (mineral water)  ...  0.013604    1.122357
2  (mineral water)           (eggs)  ...  0.008090    1.043158
3           (eggs)  (mineral water)  ...  0.008090    1.062815
4  (mineral water)    (ground beef)  ...  0.017507    1.088672
5    (ground beef)  (mineral water)  ...  0.017507    1.305401
6  (mineral water)           (milk)  ...  0.017105    1.089850
7           (milk)  (mineral water)  ...  0.017105    1.209650
8  (mineral water)      (spaghetti)  ...  0.018223    1.102008
9      (spaghetti)  (mineral water)  ...  0.018223    1.159314

[10 rows x 9 columns]

So this means when the customer will buy one of those items: chocolate,eggs,ground beef,milk or spaghetti he will definitly buy mineral water with it. 